_posts/2023-11-29-figgy-migration-details.md

```diff
@@ -0,0 +1,153 @@
+psql -d <database_name> -U <postgres_admin_user> -f /tmp/<database_name>-schema.sql
+```
+- Create a subscription on the new PostgreSQL 15 database so it will listen for updates from the PostgreSQL 10 database:
+``` psql CLI
+CREATE SUBSCRIPTION <project_name>_subscription
+CONNECTION 'host=<publisher_IP> port=5432 dbname=<database_name> user=<postgres_admin_user> password=<postgres_admin_password>' PUBLICATION <project_name>_publication WITH copy_data=true;
+```
+Once everything is set up, wait for full replication to happen. You can validate that replication is complete by generating row counts in both databases and comparing those numbers:
+``` psql CLI
+WITH tbl AS
+  (SELECT table_schema,
+          TABLE_NAME
+   FROM information_schema.tables
+   WHERE TABLE_NAME not like 'pg_%'
+     AND table_schema in ('public'))
+SELECT table_schema,
+       TABLE_NAME,
+       (xpath('/row/c/text()', query_to_xml(format('select count(*) as c from %I.%I', table_schema, TABLE_NAME), FALSE, TRUE, '')))[1]::text::int AS rows_n
+FROM tbl
+ORDER BY rows_n DESC;
+```
+
+Once replication was complete, we wanted to be sure our upgraded database had resilience built in, so we set up a warm standby in addition to our usual backup process.
+
+### How to Migrate the Application
+
+You will want to modify the process below to suit your local environment. Make each step specific so you can copy and paste.
+
+
+This is how we did it:
+
+- Select a page to use for testing. Make sure it loads and save the URL.
+- Open a terminal to each database server:
+    * `ssh user@old-database`
+    * `ssh user@new-database`
+- Open a SQL terminal connected to `<database_name>` on both servers: `sudo -u postgres psql -d <database_name>`
+- Check row counts for all tables between both servers:
+```sql
+WITH tbl AS
+  (SELECT table_schema,
+          TABLE_NAME
+   FROM information_schema.tables
+   WHERE TABLE_NAME not like 'pg_%'
+     AND table_schema in ('public'))
+SELECT table_schema,
+       TABLE_NAME,
+       (xpath('/row/c/text()', query_to_xml(format('select count(*) as c from %I.%I', table_schema, TABLE_NAME), FALSE, TRUE, '')))[1]::text::int AS rows_n
+FROM tbl
+ORDER BY rows_n DESC;
+```
+- Assuming they're close, move on. If they're far off, stop here. Something's broken.
+- Notify your users. We did this using Slack: `@here We're now starting scheduled maintenance of Figgy, it will be going into read only mode for up to 3 hours. We'll send another message here when it's ready for regular use again.`
+- Set your application to read-only mode (or shut it down), so nobody can add new records while you migrate.
+- Check row counts for all tables between both servers again (belt and suspenders, folks).
+- If they're not exact, wait a minute, try again. If they never become exact, then stop. Otherwise, continue.
+- [Update the sequences](https://wiki.postgresql.org/wiki/Fixing_Sequences) in the new database.
+- Run the following query in the new database's psql terminal:
```

Comment:
We give the link but then copy its contents below. I suggest just having one line like
```suggestion
- Run the following query (which we pulled from the [postgresl.org wiki](https://wiki.postgresql.org/wiki/Fixing_Sequences)) in the new database's psql terminal to update the sequences:
```